src/peer_connection.cpp

```diff
@@ -605,7 +605,8 @@ namespace libtorrent
 			for (int i = 0; i < int(hash.size() / sizeof(std::uint32_t)); ++i)
 			{
 				++loops;
-				piece_index_t const piece(detail::read_uint32(p) % num_pieces);
+				TORRENT_ASSERT(num_pieces > 0);
+				piece_index_t const piece(int(detail::read_uint32(p) % std::uint32_t(num_pieces)));
```

Comment:
could you use ``read_int32()`` instead, and not need the cast?
modulo should still yield a negative value if the input is negative, it it won't do any harm (as far as I can tell)